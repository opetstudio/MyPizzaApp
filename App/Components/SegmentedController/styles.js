import {Colors as colors} from '../../Themes'

export const styles = {
  tabBarContainer: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'flex-start',
    justifyContent: 'center',
    marginBottom: 20
  },
  tabButtonContainer: {
    flex: 1,
    borderBottomWidth: 2,
    borderBottomColor: 'transparent',
    backgroundColor: colors.darkGreyPrimary,
    height: 48,
    justifyContent: 'center',
    marginLeft: -1
  },
  tabSelectedButtonContainer: {
    borderBottomColor: colors.redPrimary
  }
}
