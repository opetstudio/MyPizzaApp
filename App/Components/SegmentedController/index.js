import React, { Component } from 'react'
import { View } from 'react-native'
import PropTypes from 'prop-types'
import { styles } from './styles'
import Button from '../Button'

const propTypes = {
  tabs: PropTypes.arrayOf(
    PropTypes.shape({
      i18nKey: PropTypes.string.isRequired,
      id: PropTypes.string.isRequired,
      content: PropTypes.object.isRequired
    })
  ),
  selectedId: PropTypes.string.isRequired,
  upperCase: PropTypes.bool,
  selectedStyle: PropTypes.object,
  style: PropTypes.object
}

class SegmentedController extends Component {
  state = {
    selectedId: this.props.selectedId
  };

  componentWillReceiveProps (nextProps) {
    if (this.props.selectedId !== nextProps.selectedId) {
      this.setState({ selectedId: nextProps.selectedId })
    }
  }

  renderTabs = () => {
    // ensure to use arrow function to auto bind 'this'
    const { tabs, selectedStyle, style, upperCase } = this.props
    const { selectedId } = this.state
    return (
      <View style={styles.tabBarContainer}>
        {tabs.map(tab => {
          const { id, i18nKey } = tab
          const normalStyle = [styles.tabButtonContainer, style]
          const fullSelectedStyle = normalStyle.concat([
            styles.tabSelectedButtonContainer,
            selectedStyle
          ])
          return (
            <View
              key={id}
              style={id === selectedId ? fullSelectedStyle : normalStyle}
            >
              <Button
                onPress={() => this.setState({ selectedId: id })}
                i18nKey={i18nKey}
                upperCase={upperCase}
                type='tab'
              />
            </View>
          )
        })}
      </View>
    )
  };

  renderContent = () => {
    const { tabs } = this.props
    const { selectedId } = this.state
    return (
      <View>
        {tabs.map(tab => {
          const { id, content } = tab
          return (
            <View
              key={id}
              style={{ display: id === selectedId ? null : 'none' }}
            >
              {content}
            </View>
          )
        })}
      </View>
    )
  };

  render () {
    return (
      <View>
        {this.renderTabs()}
        {this.renderContent()}
      </View>
    )
  }
}

SegmentedController.propTypes = propTypes

export default SegmentedController
