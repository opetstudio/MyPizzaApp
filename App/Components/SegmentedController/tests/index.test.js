import React from 'react';
import { shallow } from 'enzyme';
import { View } from 'react-native';
import SegmentedController from '../index';

let wrapper;

beforeEach(() => {
  wrapper = shallow(
    <SegmentedController
      tabs={[
        {
          i18nKey: 'tab-1-title',
          id: 'tab-1',
          content: <View id={'tab-1-content'} />,
        },
        {
          i18nKey: 'tab-2-title',
          id: 'tab-2',
          content: <View id={'tab-2-content'} />,
        },
      ]}
      selectedId={'tab-1'}
    />,
  );
});

it('renders SegmentedController correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

it('has two tabs', () => {
  const tabs = wrapper.findWhere(n => {
    return n.prop('i18nKey') && n.prop('i18nKey').match(/^tab-.-title$/);
  });
  expect(tabs).toHaveLength(2);
});

it('has two content segments', () => {
  const segments = wrapper.findWhere(n => {
    return n.prop('id') && n.prop('id').match(/^tab-.-content$/);
  });
  expect(segments).toHaveLength(2);
});

it('shows only tab 1 and hides tab 2', () => {
  const segmentOne = wrapper.findWhere(n => {
    return n.prop('id') && n.prop('id').match(/^tab-1-content$/);
  });
  const segmentTwo = wrapper.findWhere(n => {
    return n.prop('id') && n.prop('id').match(/^tab-2-content$/);
  });
  expect(segmentOne.parent().prop('style').display).not.toBe('none');
  expect(segmentTwo.parent().prop('style').display).toBe('none');
});
