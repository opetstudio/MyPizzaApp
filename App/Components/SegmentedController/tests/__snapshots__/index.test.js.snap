// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SegmentedController correctly 1`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "flex-start",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginBottom": 20,
      }
    }
  >
    <View
      key="tab-1"
      style={
        Array [
          Object {
            "backgroundColor": "#0f2236",
            "borderBottomColor": "transparent",
            "borderBottomWidth": 2,
            "flex": 1,
            "height": 48,
            "justifyContent": "center",
          },
          undefined,
          Object {
            "borderBottomColor": "#e5131d",
          },
          undefined,
        ]
      }
    >
      <Button
        i18nKey="tab-1-title"
        onPress={[Function]}
      />
    </View>
    <View
      key="tab-2"
      style={
        Array [
          Object {
            "backgroundColor": "#0f2236",
            "borderBottomColor": "transparent",
            "borderBottomWidth": 2,
            "flex": 1,
            "height": 48,
            "justifyContent": "center",
          },
          undefined,
        ]
      }
    >
      <Button
        i18nKey="tab-2-title"
        onPress={[Function]}
      />
    </View>
  </View>
  <View>
    <View
      key="tab-1"
      style={
        Object {
          "display": null,
        }
      }
    >
      <View
        id="tab-1-content"
      />
    </View>
    <View
      key="tab-2"
      style={
        Object {
          "display": "none",
        }
      }
    >
      <View
        id="tab-2-content"
      />
    </View>
  </View>
</View>
`;
